src/aleph/vm/controllers/firecracker/executable.py

```diff
@@ -251,8 +260,37 @@ async def wait_for_init(self) -> None:
         May be empty."""
         return
 
+    def save_controller_configuration(self):
+        """Save VM configuration to be used by the controller service"""
+        with open(f"{settings.EXECUTION_ROOT}/{self.vm_hash}-controller.json", "wb") as controller_config_file:
+            controller_config_file.write(
+                self.controller_configuration.json(by_alias=True, exclude_none=True, indent=4).encode()
+            )
+            controller_config_file.flush()
+            config_file_path = Path(controller_config_file.name)
+            config_file_path.chmod(0o644)
+            return config_file_path
+
```

Comment:
This can be simplified

```suggestion
    def save_controller_configuration(self) -> Path:
        """Save VM configuration to be used by the controller service"""
        config_file_path = Path(f"{settings.EXECUTION_ROOT}/{self.vm_hash}-controller.json")
        with config_file_path.open("w") as controller_config_file:
            controller_config_file.write(
                self.controller_configuration.json(by_alias=True, exclude_none=True, indent=4)
            )
        config_file_path.chmod(0o644)
        return config_file_path
```

Changes made:

1.  Simplified File Handling: The Path object is created at the beginning of the function, and file opening is done with the open method of the Path object. This makes the handling of the file path and file object more straightforward.

2.    Removed Unnecessary Encoding and Flushing: Since the file is opened in text mode ("w"), there's no need to encode the JSON string manually. Also, the flush method is not explicitly required here, as closing the file (which is automatically done by the with block) ensures that the data is written to the file.

3.    Streamlined Permission Setting: The file permissions are set after the file is closed. This separation of concerns makes the code easier to understand.

Reply:
Also, why is this a method of the class and not an independent function ?

Reply:
Because it depends on the Instance/Program controller configuration and the `vm_hash`. We can create an independent method, but anyway we will need to pass the desired config. Also, it is only used internally to save the configuration, so it's better to have it inside the class.

Reply:
in general if you want to just dump the content into a file I would say you don't even need to bother using a With and everything you can just do

```
        path = Path(f"{settings.EXECUTION_ROOT}/{self.vm_hash}-controller.json")
        path.open("w").write(self.controller_configuration.json(by_alias=True, exclude_none=True, indent=4))
        path.chmod(0o644)
        return path
```